File: src/python/utils/layers/dense.py
import numpy as np
import numba

from utils.cuda.cuda import *


class Layer:
    def __init__(self, input_size, output_size, activation='relu'):
        self.input_size = input_size
        self.output_size = output_size
        self.weights = np.random.randn(output_size, input_size) * np.sqrt(2. / input_size)
        self.weight = self.weights.astype(np.float32)
        self.biases = np.zeros((output_size, 1), dtype=np.float32)
        
        # Predefine activation and gradient functions
        self._init_activation(activation)

    def _init_activation(self, activation):
        if activation == 'relu':
            self.activation = self.relu
            self.activation_grad = self.relu_grad
        elif activation == 'sigmoid':
            self.activation = self.sigmoid
            self.activation_grad = self.sigmoid_grad
        elif activation == 'tanh':
            self.activation = self.tanh
            self.activation_grad = self.tanh_grad
        else:
            self.activation = self.linear
            self.activation_grad = self.linear_grad

    def forward(self, x):
        if x.ndim == 1:
            x = x.reshape(-1, 1)
        self.input = x
        self.z = np.dot(self.weights, x) + self.biases
        return self.activation(self.z)

    def backward(self, grad_output):
        # Ensure grad_output is 2D for matrix operations
        if grad_output.ndim == 1:
            grad_output = grad_output.reshape(-1, 1)

        # Calculate the gradient of the activation function
        grad_z = grad_output * self.activation_grad(self.z)

        # Calculate gradients for weights and biases
        self.grad_weights = np.dot(grad_z, self.input.T)
        self.grad_biases = grad_z.sum(axis=1, keepdims=True)

        # Propagate the gradient backwards
        grad_input = np.dot(self.weights.T, grad_z)
        return grad_input


    # Activation functions and their gradients
    @staticmethod
    @numba.njit
    def relu(z):
        return np.maximum(0, z)

    @staticmethod
    @numba.njit
    def relu_grad(z):
        return np.where(z > 0, 1.0, 0.0)

    @staticmethod
    @numba.njit
    def sigmoid(z):
        return 1 / (1 + np.exp(-z))

    @staticmethod
    @numba.njit
    def sigmoid_grad(z):
        sig = 1 / (1 + np.exp(-z))
        return sig * (1 - sig)

    @staticmethod
    @numba.njit
    def tanh(z):
        return np.tanh(z)

    @staticmethod
    @numba.njit
    def tanh_grad(z):
        return 1 - np.tanh(z) ** 2

    @staticmethod
    @numba.njit
    def linear(z):
        return z

    @staticmethod
    @numba.njit
    def linear_grad(z):
        return np.ones_like(z)


class GPULayer(Layer):
    def __init__(self, input_size, output_size, activation='relu'):
        super().__init__(input_size, output_size, activation)

        self.activation_type = activation

    def forward(self, x):
        if x.ndim == 1:
            x = x.reshape(-1, 1)
        self.input = x
        self.z = gpu_matrix_vector_mul(self.weights, x)
        self.z += self.biases
        return gpu_activation(self.activation_type, self.z, self.z.size)

    def backward(self, grad_output):
        if grad_output.ndim == 1:
            grad_output = grad_output.reshape(-1, 1)
        grad_z = gpu_elementwise_mul(grad_output, self.activation_grad(self.z))
        self.grad_weights = gpu_matrix_vector_mul(grad_z, self.input.T)
        self.grad_biases = gpu_sum(grad_z)
        grad_input = gpu_matrix_vector_mul(self.weights.T, grad_z)
        return grad_input

File: src/rust/src/layers/dense.rs
use ndarray::prelude::*;
use ndarray_rand::rand_distr::Uniform;
use ndarray_rand::RandomExt;
use std::any::Any;

use super::Layer;

pub struct DenseLayer {
    input_size: usize,
    output_size: usize,
    weights: Array2<f64>,
    biases: Array2<f64>,
    activation: Activation,
    input: Option<Array2<f64>>,
    z: Option<Array2<f64>>,
    grad_weights: Option<Array2<f64>>,
    grad_biases: Option<Array2<f64>>,
}

impl DenseLayer {
    pub fn new(input_size: usize, output_size: usize, activation: Option<String>) -> Self {
        let weights = Array2::random((output_size, input_size), Uniform::new(-1.0, 1.0)) * (2.0 / input_size as f64).sqrt();
        let biases = Array2::zeros((output_size, 1));
        let activation = match activation.as_deref() {
            Some("relu") => Activation::ReLU,
            Some("sigmoid") => Activation::Sigmoid,
            Some("tanh") => Activation::Tanh,
            _ => Activation::Linear,
        };

        DenseLayer {
            input_size,
            output_size,
            weights,
            biases,
            activation,
            input: None,
            z: None,
            grad_weights: None,
            grad_biases: None,
        }
    }

    fn activation(&self, z: &Array2<f64>) -> Array2<f64> {
        match self.activation {
            Activation::ReLU => z.mapv(|v| if v > 0.0 { v } else { 0.0 }),
            Activation::Sigmoid => z.mapv(|v| 1.0 / (1.0 + (-v).exp())),
            Activation::Tanh => z.mapv(|v| v.tanh()),
            Activation::Linear => z.clone(),
        }
    }

    fn activation_grad(&self, z: &Array2<f64>) -> Array2<f64> {
        match self.activation {
            Activation::ReLU => z.mapv(|v| if v > 0.0 { 1.0 } else { 0.0 }),
            Activation::Sigmoid => {
                let sig = z.mapv(|v| 1.0 / (1.0 + (-v).exp()));
                sig * (1.0 - sig)
            }
            Activation::Tanh => z.mapv(|v| 1.0 - v.tanh().powi(2)),
            Activation::Linear => Array2::ones(z.raw_dim()),
        }
    }
}

impl Layer for DenseLayer {
    fn as_any(&self) -> &dyn Any {
        self
    }

    fn forward(&mut self, x: &Array2<f64>) -> Array2<f64> {
        let x = if x.ndim() == 1 {
            x.clone().insert_axis(Axis(1)).into_dimensionality::<Ix2>().unwrap()
        } else {
            x.clone()
        };
        self.input = Some(x.clone());
        self.z = Some(self.weights.dot(&x) + &self.biases);
        self.activation(&self.z.as_ref().unwrap())
    }    

    fn backward(&mut self, grad_output: &Array2<f64>) -> Array2<f64> {
        let grad_output = if grad_output.ndim() == 1 {
            grad_output.clone().insert_axis(Axis(1)).into_dimensionality::<Ix2>().unwrap()
        } else {
            grad_output.clone()
        };
    
        let grad_z = grad_output * self.activation_grad(&self.z.as_ref().unwrap());
    
        self.grad_weights = Some(grad_z.dot(&self.input.as_ref().unwrap().t()));
        self.grad_biases = Some(grad_z.sum_axis(Axis(1)).insert_axis(Axis(1)));
    
        self.weights.t().dot(&grad_z)
    }    

    fn update_weights(&mut self, optimizer: &mut dyn crate::optimizer::Optimizer, layer_index: usize) {
        self.weights = optimizer.update(
            &self.weights,
            self.grad_weights.as_ref().unwrap(),
            &format!("layer_{}_weights", layer_index),
        );
        self.biases = optimizer.update(
            &self.biases,
            self.grad_biases.as_ref().unwrap(),
            &format!("layer_{}_biases", layer_index),
        );
    }

    fn output_size(&self) -> usize {
        self.output_size
    }
}

enum Activation {
    ReLU,
    Sigmoid,
    Tanh,
    Linear,
}
